File: 4.py
#!/usr/bin/env/python

# Find largest palindrome made from the product of two 3-digit numbers

# doing a reverse cycle, from 999 and back

def is_palindrom(n):
    " Check is a number is palindrom"
    if list(str(n)) == list(reversed(str(n))):
        return True
    return False


adders = [999, 999]
idx = 0

while True:
    #print "adders now %d %d" % (adders[0], adders[1])
    prod = adders[0] * adders[1]
    if is_palindrom(prod):
        print prod
        break
    
    adders[idx] -= 1
    idx = (idx + 1) % 2
        

File: 4.hs
--Find largest palindrome made from the product of two 3-digit numbers
module Main where

is_pal n = show n == reverse (show n)

-- brute force, using commutative property of *
gpal = filter is_pal [x * y | x <- [100..999], y <- [x..999]]
main = do
  putStrLn $ show $ foldr1 max gpal
